httpclient5/src/main/java/org/apache/hc/client5/http/impl/routing/DistributedProxySelector.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+package org.apache.hc.client5.http.impl.routing;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.SocketAddress;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A ProxySelector implementation that selects proxies using a distributed strategy.
+ * <p>
+ * This implementation maintains a list of other ProxySelectors, and selects a proxy
+ * by delegating to the next available ProxySelector in the list, until a non-empty
+ * proxy list is returned. If a ProxySelector fails with an exception, the next
+ * ProxySelector in the list is tried. Once a non-empty proxy list is returned by
+ * a ProxySelector, it becomes the active ProxySelector and is used for subsequent
+ * calls to select(). If a connection fails, connectFailed() is called on the
+ * active ProxySelector.
+ * <p>
+ * This implementation is useful in environments where multiple proxy servers may be
+ * used, and where it is desirable to distribute proxy selection among different
+ * ProxySelectors.
+ * <p>
+ * This implementation uses a ThreadLocal to store the currently chosen selector
+ * for each thread, ensuring thread-safety. Additionally, it uses a cache to store
+ * previously selected proxies for each URI, improving performance by avoiding
+ * redundant proxy selection calls.
+ *
+ * @since 5.3
+ */
+public class DistributedProxySelector extends ProxySelector {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DistributedProxySelector.class);
+
+    private final List<ProxySelector> selectors;
+    private final ThreadLocal<Integer> currentIndex;
```

Comment:
@arturobernalg 
1. You can stick `ProxySelector` into `ThreadLocal`. It will make the logic more readable.
2. Do not forget to remove it in `#connectFailed`

Reply:
@ok2c 
I have now incorporated this change into the DistributedProxySelector class, which has made the logic more readable and ensures thread safety.

The updated implementation now utilizes ThreadLocal<ProxySelector> for the currentSelector, allowing each thread to have its own instance of the ProxySelector. Additionally, I have updated the connectFailed method to remove the currentSelector from the ThreadLocal variable after handling the connection failure.